aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -756,9 +785,97 @@ Tensor dense_to_sparse_bsr(const Tensor& self, IntArrayRef blocksize) {
+
+  if (n_batch_dim > 1) {
+    // Flatten to single batch dim
+    values = values.flatten(0, n_batch_dim - 1);
+    not_zero_mask = not_zero_mask.flatten(0, n_batch_dim - 1);
+  }
+  if (is_batched) {
+    // Some extra setup and checks for batched
+    TORCH_CHECK(
+        not_zero_mask.size(0) > 0,
+        "to_sparse_bsc: Expected product of batch dimensions to be non-zero.");
+
+    // Batched inputs must have the  same nnz per batch.
+    auto not_zero_mask_0 = not_zero_mask.select(0, 0);
+    auto nse_per_batch = not_zero_mask_0.sum().repeat(not_zero_mask.size(0));
+    TORCH_CHECK(
+        not_zero_mask.sum({-2, -1}).equal(nse_per_batch),
+        "Expect the same number of specified elements per batch.");
+
+    // We need to join batches into a matrix increasing the number of cols.
+    // This allows us to create indices for col compressed matrix and de-batch
+    // them later. Avoiding a kernel start for each batch.
+    values = values.transpose(0, 1).flatten(1, 2);
+    not_zero_mask = not_zero_mask.transpose(0, 1).flatten(1, 2);
+  }
+
+  Tensor col_indices;
+  Tensor row_indices;
+  // Compressed col indices are the same as the row indices of the transpose!
+  std::tie(row_indices, col_indices) = _not_zero_mask_to_col_row_indices(
+      not_zero_mask.transpose(1, 0), at::kLong, not_zero_mask.device());
+  // This only works if the col_indices vector is in ascending order.
+  Tensor ccol_indices = at::_convert_indices_from_coo_to_csr(
+      col_indices, not_zero_mask.size(-1), false /*out_int32*/);
+  // We need the values to blocked row major (they are), but blocks to be col
+  // major, so we transpose the leading two dims The mask must transpose as well
+  // to index it correctly
+  values = values.transpose(0, 1).flatten(0, -3).index_select(
+      0, _mask_to_indices(not_zero_mask.transpose(0, 1).flatten()));
+  if (is_batched) {
+    auto batch_shape = self.sizes().slice(0, n_batch_dim);
+    auto n_batch = std::accumulate(
+        batch_shape.begin(),
+        batch_shape.end(),
+        int64_t{1},
+        std::multiplies<int64_t>());
+    // Note nonzero_mask.size(-1) / block_size_1 leads to FPE when cols is zero
+    ccol_indices = at::_compressed_to_batched_compressed_indices(
+        ccol_indices, n_batch, false /*out_int32*/);
+
+    auto batchsize_infer_last = DimVector(batch_shape);
+    // we can infer nnz/ncol+1
+    batchsize_infer_last.push_back(-1);
+    // -1 dim is nnz (per batch)
+    row_indices = row_indices.reshape(batchsize_infer_last);
+    // -1 dim is nnrows (per batch) + 1
+    ccol_indices = ccol_indices.reshape(batchsize_infer_last);
+    // -1 dim is nnz (per batch) blocksize dims already included (hence
+    // unflatten)
+    values = values.unflatten(0, batchsize_infer_last);
+  }
+
+  return at::native::_sparse_bsc_tensor_unsafe(
+      ccol_indices,
+      row_indices,
+      values,
+      self.sizes(),
+      values.scalar_type(),
+      c10::kSparseBsc,
```

Comment:
curious what `_sparse_bsc_tensor_unsafe` does with this argument 😁

Reply:
I think it just checks that it is exactly this value? I know it is required, but it probably does not have to be. I think it stems from a common factory function which is the underlying implementation for the generic `_sparse_compresse_tensor_unsafe` as well as each of the specialized `_sparse_<XXX>_tensor_unsafe`